### PR TITLE
Kqueue backend for use on BSD

### PR DESCRIPTION
**Pull request by xanderio**
Kqueue Backend for all the BSDs 🎉 

It's based on the kqueue crate and the in watcher implementation is based on the `INotifyWatcher`.

BSDs where this PR is check that it compiles: 

- [x] FreeBSD
- [ ] OpenBSD
- [ ] NetBSD
- [ ] DragonflyBSD

This relates to #327 and maybe #136

### CODE REVIEW

**Comment by 0xpr03** [comment]
I enabled the CI for this PR so if possible you can setup & test github CI for freebsd

**Comment by xanderio** [comment]
Thanks, but Github Actions don't have support for FreeBSD.

I found this action[1], but i'm not sure if this is a greate solution for that problem.

[1]: https://github.com/vmactions/freebsd-vm

**Comment by xanderio** [comment]
We could setup a ci pipeline under MacOS in combination with a feature to select between FSEvent and kqueue.

**Comment by JohnTitor** [comment]
FWIW, the libc crate and Tokio use Cirrus CI (e.g. https://github.com/tokio-rs/tokio/blob/master/.cirrus.yml) for FreeBSD testing though it has some limitations for free users.

**Comment by xanderio** [comment]
Which solution should i implement? 

1. Cargo feature to toggle the mac backend to kqueue
2. Cirrus CI on FreeBSD

**Comment by xanderio** [comment]
The only thing the I'm currently seeing that is still missing is some kind of CI solution.

**Comment by JohnTitor** [comment]
I don't have a strong opinion here but is the first option beneficial for mac users as well? If so, it's better than the second as we don't need additional CI.

**Comment by 0xpr03** [comment]
I think we want some freebsd CI if possible. I'm fine with testing this only on mac if nothing else, but to have official BSD support it'd be very nice to have a CI run. Otherwise you have a very high chance it'll break and nobody will notice. @JohnTitor can you help setup cirrus CI for that if possible ?

**Comment by JohnTitor** [comment]
> Otherwise you have a very high chance it'll break and nobody will notice.

AFAIK macOS is one of BSD targets and it should have kqueue API. Does this have FreeBSD-specific operations?

I'm happy to help setup Cirrus CI though :)

**Comment by worr** [comment]
Hey, original author of the `kqueue` crate here.

This doesn't have anything FreeBSD-specific, since as of right now, the high-level parts of the kqueue are all cross-platform. I've looked at the NOTEs and stuff they're using and ensured that they're also cross-platform. In other words, LGTM.

While macOS does support `kqueue(2)` and theoretically this could be used there, not sure what cases someone would want to use `kqueue(2)` rather than `FSEvents` which is more native, and afaict already supported by this crate. A feature like @xanderio suggested might be a good way to gate it, for either easier CI or in case someone v specifically wants to use `kqueue(2)`. That's just my 2¢ tho, feel free to disregard 🙂

File: src/kqueue.rs
Comment (0xpr03) [inline src/kqueue.rs:142]: I think this should be removed before merging

File: src/kqueue.rs
Comment (0xpr03) [inline src/kqueue.rs:308]: The caller functions returns an error, so I'd like to report permission/access errors here too and not silently drop them

File: src/kqueue.rs
Comment (0xpr03) [inline src/kqueue.rs:337]: I think we could also just report this via the Result and not panic randomly inside application code. But that's not a hard opinion, just an idea.

File: src/kqueue.rs
Comment (0xpr03) [inline src/kqueue.rs:353]: same goes here if applicable

File: src/kqueue.rs
Comment (0xpr03) [inline src/kqueue.rs:166]: *written

File: src/kqueue.rs
Comment (0xpr03) [inline src/kqueue.rs:176]: You can change this to `kqueue::Vnode::Extend | kqueue::Vnode::Truncate =>` ..

File: src/kqueue.rs
Comment (0xpr03) [inline src/kqueue.rs:196]: *no idea

File: src/kqueue.rs
Comment (0xpr03) [inline src/kqueue.rs:205]: @worr do you know of a better way than to re-index the whole tree ?

**Review by 0xpr03 — CHANGES_REQUESTED** [review]
Thanks for the PR. I've added some minor questions or notes what could be improved. If possible I'd like to resolve the todos. I think we can also just track some stuff if we're already re-indexing the whole path upon link change.

File: src/kqueue.rs
Comment (xanderio) [reply src/kqueue.rs:308]: This code is copied directly from the INotifyWatcher implementation. So we should resolve this also in there

File: src/kqueue.rs
Comment (xanderio) [reply src/kqueue.rs:353]: This is also copied from the INotifyWatcher

File: src/kqueue.rs
Comment (xanderio) [reply src/kqueue.rs:337]: This is also copied the INotifyWatcher

File: src/kqueue.rs
Comment (xanderio) [reply src/kqueue.rs:201]: We could store a HashMap of all know paths and then compare with paths are missing or where added to the directory.
I'll try to implement this.

Comment (xanderio) [reply src/kqueue.rs:201]: After thinking about this a bit, i cam to the conclusion that a simple `HashSet<PathBuf>` won't cut it. As i don't think there is any way to get from the directory we get from `NOTE_LINK` to the files we expect to be in that directory.

I think we need a `HashMap<PathBuf, Vec<PathBuf>>` where the key is the directory and the value is a list of all files in the directory at time of adding. This feels like a race condition to me, as files could be created of deleted will we are setting up this directory watch. Not quite sure if this is important in this case.

File: src/kqueue.rs
Comment (xanderio) [reply src/kqueue.rs:333]: After looking over this peace of code, i agree with you. To which `ErrorKind` should i map this, or should this maybe be resolved in an another  PR where this behavior is also removed from `INotifyWatcher`.

File: src/kqueue.rs
Comment (xanderio) [reply src/kqueue.rs:349]: After looking over this peace of code, i agree with you. To which `ErrorKind` should i map this, or should this maybe be resolved in an another PR where this behavior is also removed from `INotifyWatcher`.

File: src/kqueue.rs
Comment (worr) [reply src/kqueue.rs:201]: @0xpr03 We've been chatting about this out-of-band a bit. `kqueue(2)` doesn't expose this in a meaningful way, and this is especially expensive since we need to open file handles to each file on index (this is a requirement of `kqueue(2)`). The method @xanderio brought up is one of the ones we've been talking about.

Comment (0xpr03) [reply src/kqueue.rs:201]: Hm, thanks for investigating. I'm fine with the current solution, it's not the first time where we have to re-index due to race conditions. Could you change the todo so it documents why we have to use this approach ?

File: src/kqueue.rs
Comment (0xpr03) [reply src/kqueue.rs:349]: error::Generic as also requested [here](https://github.com/notify-rs/notify/pull/337#discussion_r652165310) for anther in-flight PR

File: src/kqueue.rs
Comment (0xpr03) [reply src/kqueue.rs:333]: error::Generic is fine

**Comment by 0xpr03** [comment]
Ping me again here if you're not hearing anything from me, I just discovered your comments on the review by accident and didn't get any notification. (Also the github CI run apparently has to be re-approved every time again, as long as you haven't made at least one commit into the repo.)

File: src/kqueue.rs
Comment (xanderio) [reply src/kqueue.rs:201]: I'll can do that, all though I personally would prefer if  we left the TODO in places where there are thinks that we see that can possibly be improve in the future, but we just don't know the solution to at the moment. But that's just my 2ct

**Comment by xanderio** [comment]
Quick side note: After this is merged could you release a new preview build?

**Comment by 0xpr03** [comment]
Yeah, I'll probably merge #337 first and then release them together when I cleaned up the changelog, some stuff got left out.

**Comment by xanderio** [comment]
After we fix the could of code review issues with this pr. The only thing that is missing is some kind of CI setup. 
I would start working on the macos based solution previously discussed, if i don't here anything else from you.

**Comment by 0xpr03** [comment]
I'm sorry but I'm not exactly sure what you mean with that. If you want you can go ahead and make a macOS PR.

File: src/kqueue.rs
Comment (0xpr03) [reply src/kqueue.rs:201]: on a second thought yeah, you can leave it like that

**Comment by xanderio** [comment]
I mean adding a cargo `macos_kqueue` feature which enables the `kqueue` backend for macos. With that we could add CI to this PR without having to figure out how to setup cirrus ci for an opensource project.

**Comment by 0xpr03** [comment]
Yeah that sounds like a good idea. @JohnTitor do you prefer doing this via cirrus or just a cargo feature ?

**Comment by xanderio** [comment]
I would like to leave the  `TODO` on the `NOTE_DELETE` in the source code, because this is a valid improvement that can be made in the future.

File: src/kqueue.rs
Comment (0xpr03) [inline src/kqueue.rs:11]: HashSet should be irrelevant then

**Review by 0xpr03 — CHANGES_REQUESTED** [review]

**Comment by 0xpr03** [comment]
> 
> 
> I would like to leave the `TODO` on the `NOTE_DELETE` in the source code, because this is a valid improvement that can be made in the future.

ok

**Comment by xanderio** [comment]
The latest commit include a CI pipeline, as this is the first time i ever did something with github actions, I'm not sure if this is correct. 

Other then the unwrap issue on line 343 and 359. This PR is in my eyes almost complete. 🎉

File: .github/workflows/main.yml
Comment (JohnTitor) [inline .github/workflows/main.yml:45]: IIRC we can use `${{ matrix.feature }}` directly.

**Comment by JohnTitor** [comment]
> Yeah that sounds like a good idea. @JohnTitor do you prefer doing this via cirrus or just a cargo feature ?

I don't have a strong opinion but it's worth trying out the latter if we've implemented it.

File: .github/workflows/main.yml
Comment (JohnTitor) [inline .github/workflows/main.yml:25]: > The workflow is not valid. .github/workflows/main.yml (Line: 25, Col: 18): Matrix vector 'feature' does not contain any values

The default could be "serde" so that we can simplify the below command on `check build`.

Comment (JohnTitor) [reply .github/workflows/main.yml:25]: nit: And it should be `features`.

**Comment by xanderio** [comment]
Just pushed to last commit, which should ™️  fix all code review issus

File: .github/workflows/main.yml
Comment (JohnTitor) [inline .github/workflows/main.yml:42]:
```suggestion
```

**Comment by xanderio** [comment]
Thanks for fixing the CI @JohnTitor 

@0xpr03 According to github you're still requesting changes, but i can't figure out what changes those are. :/

**Review by 0xpr03 — APPROVED** [review]

**Comment by 0xpr03** [comment]
I think that was simply because I had a change-request review with a [general comment](https://github.com/notify-rs/notify/pull/335#pullrequestreview-683781485) which probably isn't resolved by the individual code comments.

**Comment by xanderio** [comment]
I'm wondering if there is something that this PR is waiting on or if you just haven't come around to merge it?

**Comment by 0xpr03** [comment]
Thanks, was busy last week and githubs UI made it looks like merging this isn't possible without manual merge conflict intervention.

**Pull request merged**
